### PR TITLE
[TG-1058] Fixes parsing errors relating inner classes on generic classes

### PR DESCRIPTION
**Pull request by thk123**
When parsing types that were inner classes inside generic classes

```java
class Outer<T> {
  class Inner { }
  void foo(Inner input) {} 
  Inner field;
}
```
Then you end up with singatures like:

```
LOuter<TT;>Inner;
```

The parsing was only parsing up the the first `;` or `<` meaning we missed out on the inner class. 

This PR resolves this (and a linked issue of if the inner class is generic parsing the generic details from within it). Most of this PR is in the unit tests, so I suggest reviewing commit by commit. I will first squash the commits a bit. Then re-run clang-format to tidy them up.

### CODE REVIEW

**Comment by thk123** [comment]
This is based on an old version of a develop (49554177f8f467ea0bbbf9185ab68739694e8013) which is the current version on TG. This should be bumped before this is merged.

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: Don't take by ref (it happens to work thanks to automagic lifetime extension, but it suggests something different to what's actually happening)

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: Kick this out into a non-member static function, it's too long for an inline lambda

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: Just return it directly?

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: Similarly, might as well return directly

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: To avoid duplcating this at line 234, turn this into `while(true)` and simply put the break condition in the middle of the loop

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp:271]: Suggest giving an example (in a comment) of a typical typename containing `.` and `/`, and how it appears afterwards

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: Typo individual

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: typo generic

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: Break line after `(`, indent params by two
e.g.
```
fghij(
  int x,
  int y)
```

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: Before these: precondition src.size()>=2

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: INVARIANT(new_type.id()!=ID_symbol)?

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: This looks very similar to `java_type_from_string`'s function parameter list parsing, can they be factored?

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: Take input string by ref?

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp:357]: Same loop rotation recommendation as above -- prefer writing this update condition once and doing `while(true) { .... if(exit_condition) break; ... }` over duplicating it like this

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: shouldn't this be declared in the header file?

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: use `!..empty()`

File: src/java_bytecode/java_types.cpp
Comment (smowton) [inline src/java_bytecode/java_types.cpp]: This might work, but I'd like to see some argument about why this exhaustively covers the type signature grammar... if this gets too tricky we should probably abandon manual parser writing and instead make a Bison or similar parse for the signature grammar (https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3.4)

This exists but is GPL'd: https://github.com/JetBrains/jdk8u_jdk/blob/master/src/share/classes/sun/reflect/generics/parser/SignatureParser.java

**Review by smowton — COMMENTED** [review]
Looks tricky. Perhaps this is nasty enough that we need a real parser generator?

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: I don't understand this, wouldn't this return `npos` if no corresponding closing `>` can be found, not if it is recursive generic?

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: why not in the header?

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: 1->one

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: please document what the loop does

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: why not an invariant with information string?

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: as above

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: isn't this an invariant for any non-atomic class file level type ?

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp:389]: would it make sense to adapt the `find_closing_delimiter` function to take a set of opening characters? Then using `find_closing_delimiters(src, end, "LT", `;`)` would return the `;` that corresponds to the first `L` and skip over any other `L`,`;` or `T`, `;` pair

File: unit/java_bytecode/java_utils_test.cpp
Comment (mgudemann) [inline unit/java_bytecode/java_utils_test.cpp:16]: including the `cpp` file here? is this the reason for the definitions in the `cpp` file instead of the header ?

**Review by mgudemann — CHANGES_REQUESTED** [review]

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: I though move semantics meant this was OK, but updated

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: My feeling is it is replacing the body of a for loop and therefore clearer to have it inline (particularly after applying below suggestions of returning directly). Can do if you disagree

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: You mean like:
```C++
  std::size_t f_pos;
  while(true)
  {
    f_pos = class_name.find('<', 1);
    if(f_pos==std::string::npos) 
    {
      break;
    }
    std::size_t e_pos = find_closing_delimiter(class_name, f_pos, '<', '>');
    if(e_pos == std::string::npos)
      throw unsupported_java_class_signature_exceptiont("recursive generic");

    // erase generic information between brackets
    class_name.erase(f_pos, e_pos - f_pos + 1);
  }
```
I think this makes the code more confusing to follow is it now not as clear what the loop is looping over.

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: Sorry messed up applying clang-format, this should now be auto-resolved.

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: Also done for other functions with similar preconditions

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: Unless I've misunderstood this comment, it is exactly that code and this is it being factored so can be used for parsing `<TT;Ljava/lang/Object;>` type strings.

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: Sorry good spot - put in as erasing a condition I didn't think was relevant but forgot to change to an invariant.

Comment (thk123) [reply src/java_bytecode/java_types.cpp]: In fact don't think it should be an invariant, but wanted to hard crash if anywhere that was true (it isn't particularly important here) so I've just removed it.

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp:357]: See above discussion, will apply both changes if you want, otherwise neither

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: Yes I completely agree, unfortunately there isn't time at this juncture to do this. Here is why I'm convinced this is reasonable for now:

This is for parsing a reference type from either a signature or descriptor. If it is descriptor, then the grammar is exactly:
`L ClassName ;`

So since ClassName can't contain `;` or `<` we're done. 

(Which is why it worked until now). 

For signatures, it is more complicated:
```
L PackageSpecifier_opt SimpleClassTypeSignature ClassTypeSignatureSuffix* ;
```

The package specifier is just `identifier/identifier` so no chances of `;` or `<'`

Then the `SimpleClassTypeSignature` is 

```
Identifier TypeArguments_opt
```
Where type arguments must start with `<` (so this is the case where `next_angle_bracket < next_semi_colon` and must end with a corresponding `>` 

This is then optionally followed by a `.` indicating inner class and then repeat above argument. It is finally ended by a semi colon. 

I appreciate this is _not_ convincing, hopefully the extensive unit tests are *more* convincing - though I accept they are a long way from an exhaustive proof.

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: This is why I prefer classes even without state, as it allows declaring in the header without making a global function of everything that includes `java_types.h` I will move the declarations to the top of the file for now and raise a ticket to refactor this parsing into a class if this is satisfactory for you?

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: As above - let me know if you'd be happy with my proposed solution

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: You are right - however this case happens precisely when we have a recursive generic type. Given ```java_generic_type_from_string(<T:LSimpleRecursiveGeneric<TT;>;>Ljava/lang/Object;)``` calls erase type arguments with `T:LSimpleRecursiveGeneric<TT`  (https://diffblue.atlassian.net/browse/TG-610)

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: Added:
```
  // Loop over the types in the given string, parsing each one in turn
  // and adding to the type_list
```

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: Removed as was only there as [previously we did](https://github.com/diffblue/cbmc/pull/1512/files#diff-ba71807a8e04db80ccb196e4eb6465feL281):

```
if(param.type().id()==ID_symbol)		
  param.type()=java_reference_type(param.type());
```

but afaict this never happens (wanted to verify this with a hard fail - but since the CI passes and I can't see a way for this to happen, I guess it is just legacy code)

File: unit/java_bytecode/java_utils_test.cpp
Comment (thk123) [reply unit/java_bytecode/java_utils_test.cpp:16]: Vice-versa, including as declarations are not in the header, not in the header to avoid making the functions global. See earlier discussion and let me know if you're happy with my proposed solution (this will still include the cpp file)

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp:389]: I think the problem is you have two types of brackets `L`, `T`, which are closed by `;` and `<` which is closed by `>`. I suppose it would in fact work as `Lfoo<;>`; isn't legal, but presumably wouldn't bec caught if that did happen. Also `LL;` would fail (class named L)

Comment (thk123) [reply src/java_bytecode/java_types.cpp:389]: And indeed the far more common `Lgeneric_class<TT;>;`

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: I thought so, changed this to an assertion and a test failed so I reverted back to original behaviour, but I can now no longer reproduce this, so perhaps I had something else wrong.

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: We don't appear to have a test case that exercises this path so I will change the exception into a more prescriptive name

**Comment by thk123** [comment]
Thanks for the reviews - I've updated PR (didn't squash commits to make it easier to review changes - but indicated in commit message what they would be squashed into). 

Will address the liner issues tomorrow but let me know about the comments I've commented on @smowton @mgudemann

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [reply src/java_bytecode/java_types.cpp]: ok with me

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [reply src/java_bytecode/java_types.cpp:389]: the idea of `find_closing_delimiters` is to support hierarchic structures (maybe it's buggy?) and allow for free definition of closing and opening delimiter.
What I imagined was to adapt it in such a way that it would in this case
 - ignore `<` and `>`
 - increase the depth at `T` and `L`
 - decrease depth at `;`

agreed for `LL;`  or `LT;`, but with one lookahead or a Boolean that indicates that the last char was an `L` and "opened" the object name, we would be good, wouldn't we?
This parsing is slightly fragile and with one common function, we'd at least know where to correct an error and don't risk overseeing another place.

**Comment by mgudemann** [comment]
@thk123 won't fight for the parsing stuff, but we should maybe think of using a parser generator and implement the grammar given in the spec.

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: Done and raised as [[TG-1206](https://diffblue.atlassian.net/browse/TG-1206)]

File: src/java_bytecode/java_types.cpp
Comment (majakusber) [reply src/java_bytecode/java_types.cpp]: Even after recursive generic types are fixed, the exception here makes sense. How about changing it to something more general like `unsupported_java_class_signature_exceptiont("Failed to find generic signature closing delimiter");` or `unsupported_java_class_signature_exceptiont("Failed to find generic signature closing delimiter (or recursive generic)");` if we want to keep the reminder that recursives cause problem here

File: unit/java_bytecode/java_utils_test.cpp
Comment (majakusber) [inline unit/java_bytecode/java_utils_test.cpp]: Should it be TV instead of VV here? (similarly in multiple places in this scenario) I thought all type variables are given prefix T.

File: src/java_bytecode/java_types.cpp
Comment (majakusber) [inline src/java_bytecode/java_types.cpp]: Same as in the other file, the exception makes sense in general, not only for recursive generics, consider changing to something like ``unsupported_java_class_signature_exceptiont("Failed to find generic signature closing delimiter");``

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp:389]: Surely if you ignore `<`, there is no difference between `LALFoo;` (a class called `ALFoo`) and `LA<LFoo;>;` and in the first case you won't find a matching `;`? 

I'll give it a go but I think this is a fudamentally different task (matching two isolated sets of brackets) and I think it's a bad idea to complicate `find_closing_delimiters`

**Comment by thk123** [comment]
@mgudemann @smowton I have raised a ticket [TG-1205](https://diffblue.atlassian.net/browse/TG-1205) to convert this conversion to using a parser generator since this logic is clearly confusing and error prone. 

I will try using `find_closing_delimiter` but other than that I think this is ready for re-review

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [reply src/java_bytecode/java_types.cpp:389]: keep it like that then, with the overall goal to get in a real parser, I am ok with this

**Review by mgudemann — APPROVED** [review]

**Review by majakusber — APPROVED** [review]
Looks good, I only had three small things - one of them you already fixed (I was looking at it commit by commit).

**Comment by thk123** [comment]
Remaining linter errors are discrepancies between clang-format and `cpplint.py` - there is an on-going discussion as to what to do with these

File: src/java_bytecode/java_types.cpp
Comment (smowton) [reply src/java_bytecode/java_types.cpp]: It's actually a lot better now that its indentation is fixed

File: unit/java_bytecode/java_utils_test.cpp
Comment (thk123) [reply unit/java_bytecode/java_utils_test.cpp]: Good spot!

**Review by smowton — APPROVED** [review]
LGTM

**Comment by thk123** [comment]
Rebased - just waiting on CI then good to merge

**Pull request merged**
